File: go/0128.longest-consecutive-sequence/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:33
// leetgo: dev
// https://leetcode.com/problems/longest-consecutive-sequence/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func longestConsecutive(nums []int) int {
	hash_map := make(map[int]bool)
	for _, val := range nums {
		hash_map[val] = true
	}

	best_count := 0
	for _, val := range nums {
		if !hash_map[val-1] {
			tmp := val
			count := 1
			for hash_map[tmp+1] {
				tmp += 1
				count += 1
			}
			best_count = max(best_count, count)
		}
		if best_count > len(nums)/2 {
			break
		}
	}
	return best_count
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	nums := Deserialize[[]int](ReadLine(stdin))
	ans := longestConsecutive(nums)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0128.longest-consecutive-sequence/solution.rs
// Created by Olgierd Palasz at 2026/07/21 19:51
// leetgo: dev
// https://leetcode.com/problems/longest-consecutive-sequence/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;
// @lc code=begin
use std::collections::{HashMap, HashSet};
impl Solution {
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let values: HashSet<i32> = nums.into_iter().collect();
        let mut seen_amount: HashMap<i32, i32> = HashMap::with_capacity(values.len());

        if values.is_empty() {
            return 0;
        }

        let mut max_ever = 1;

        for num in &values {
            // println!("num: {:?}", num);
            if !seen_amount.contains_key(num) {
                let mut count = 1;
                let mut current = *num;

                while values.contains(&(current + 1)) {
                    if let Some(value) = seen_amount.get(&(current + 1)) {
                        count += *value;
                        break;
                    } else {
                        count += 1;
                    }

                    seen_amount.insert(current + 1, 0);
                    current += 1;
                }

                seen_amount.insert(*num, count);

                max_ever = max_ever.max(count);

                // println!("num {} - {:?}", num, seen_amount);
            }
        }

        max_ever
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let ans: i32 = Solution::longest_consecutive(nums);

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
